subsys/dfu/boot/mcuboot.c

```diff
@@ -162,36 +165,89 @@ static int boot_magic_write(u32_t bank_offs)
 	return rc;
 }
 
-static int boot_magic_code_check(const u32_t *magic)
+int boot_read_bank_header(u32_t bank_offset,
+			  struct mcuboot_img_header *header,
+			  size_t header_size)
 {
-	int i;
-
-	if (memcmp(magic, boot_img_magic, sizeof(boot_img_magic)) == 0) {
-		return BOOT_MAGIC_GOOD;
+	int rc;
+	u32_t magic;
+	/*
+	 * Raw representation of the image header.
+	 *
+	 * The header_magic and header_size fields are only needed for
+	 * sanity checking that bank_offset points at a supported
+	 * image header.
+	 *
+	 * These are unsupported by MCUboot or not relevant to Zephyr now:
+	 *
+	 * - image_load_address only matters for PIC code (unsupported)
+	 * - image_flags are unsupported / currently irrelevant to Zephyr
+	 *
+	 * These are potentially useful:
+	 *
+	 * - image_size
+	 * - version
+	 */
+	struct {
+		u32_t header_magic;
+		u32_t image_load_address;
+		u16_t header_size;
+		u16_t pad;
+		u32_t image_size;
+		u32_t image_flags;
+		struct {
+			u8_t major;
+			u8_t minor;
+			u16_t revision;
+			u32_t build_num;
+		} version;
+		u32_t pad2;
+	} v1_raw;
+	struct mcuboot_img_sem_ver *sem_ver;
+	size_t v1_min_size = (sizeof(u32_t) +
+			      sizeof(struct mcuboot_img_header_v1));
+
+	/*
+	 * Only version 1 image headers are supported.
+	 */
+	if (header_size < v1_min_size) {
+		return -ENOMEM;
 	}
 
-	for (i = 0; i < ARRAY_SIZE(boot_img_magic); i++) {
-		if (magic[i] != 0xffffffff) {
-			return BOOT_MAGIC_BAD;
-		}
+	rc = flash_read(flash_dev, bank_offset, &magic, sizeof(magic));
+	if (rc) {
+		return rc;
+	} else if (magic != BOOT_HEADER_MAGIC_V1) {
+		return -EINVAL;
 	}
 
-	return BOOT_MAGIC_UNSET;
-}
-
-static int boot_magic_state_read(u32_t bank_offs)
-{
-	u32_t magic[4];
-	u32_t offs;
-	int rc;
-
-	offs = MAGIC_OFFS(bank_offs);
-	rc = flash_read(flash_dev, offs, magic, sizeof(magic));
-	if (rc != 0) {
+	/*
+	 * Read and sanity-check the raw header.
+	 */
+	rc = flash_read(flash_dev, bank_offset, &v1_raw, sizeof(v1_raw));
```

Comment:
nit - image header can be read now without reading magic.

Reply:
I guess so for v1, but it feels a bit wrong. It's safe to read `sizeof(v1_raw)` bytes from flash once we see the v1 magic number. But what if the magic number is the v2 magic number, and we shrink the header? Unlikely, but seems possible.

What do you think?